pkg/agent/controller/globalingressip.go

```diff
@@ -0,0 +1,113 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controller
+
+import (
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog"
+)
+
+const (
+	ClusterIPService           = "ClusterIPService"
+	headlessServicePod         = "HeadlessServicePod"
+	IPAllocated                = "Allocated"
+	defaultReasonIPUnavailable = "ServiceGlobalIPUnavailable"
+	defaultMsgIPUnavailable    = "Service doesn't have a global IP yet"
+)
+
+type IngressIP struct {
+	namespace   string
+	target      string
+	svcName     string
+	podName     string
+	allocatedIP string
+	condition   struct {
+		isAllocated bool
+		reason      string
+		message     string
+	}
+}
+
+func parseIngressIP(obj *unstructured.Unstructured) *IngressIP {
+	var (
+		found bool
+		err   error
+	)
+
+	gip := &IngressIP{}
+	gip.namespace = obj.GetNamespace()
+	gip.target, found, err = unstructured.NestedString(obj.Object, "spec", "target")
+	if !found || err != nil {
+		klog.Errorf("target field not found in spec %#v", obj.Object)
+		return nil
+	}
+
+	gip.svcName, found, err = unstructured.NestedString(obj.Object, "spec", "serviceRef", "name")
+	if !found || err != nil {
+		klog.Errorf("ServiceRef not found in spec %#v", obj.Object)
+		return nil
+	}
+
+	if gip.target == headlessServicePod {
+		gip.podName, found, err = unstructured.NestedString(obj.Object, "spec", "podRef", "name")
+		if !found || err != nil {
+			klog.Errorf("Expected PodRef in spec %#v", obj.Object)
+			return nil
+		}
+	}
+
+	gip.allocatedIP, _, _ = unstructured.NestedString(obj.Object, "status", "allocatedIP")
+	conditions, found, err := unstructured.NestedSlice(obj.Object, "status", "conditions")
+	if !found || err != nil || len(conditions) < 1 {
+		gip.condition.isAllocated = false
+		gip.condition.message = defaultMsgIPUnavailable
+		gip.condition.reason = defaultReasonIPUnavailable
+	} else {
+		latestCondition := conditions[len(conditions)-1].(map[string]interface{})
+		gip.condition.isAllocated = false
+		gip.condition.message = latestCondition["message"].(string)
+		gip.condition.reason = latestCondition["reason"].(string)
+		if latestCondition["type"] == IPAllocated && latestCondition["status"] == "True" {
+			gip.condition.isAllocated = true
+			if gip.allocatedIP == "" {
+				klog.Errorf("Expected AllocatedIP in status of %#v", obj.Object)
+			}
+		}
+	}
```

Comment:
As I mentioned before I think we should base the allocated status on `allocatedIP` and only consult the conditions to get the reason if not set. The conditions are really just to report info to the end user.

```suggestion
        conditions, _, _ := unstructured.NestedSlice(obj.Object, "status", "conditions")
        if gip.allocatedIP == "" {
            if len(conditions) > 0 {
                latestCondition := conditions[len(conditions)-1].(map[string]interface{})
                gip.unallocatedMessage = latestCondition["message"].(string)
                gip.unallocatedReason = latestCondition["reason"].(string)
            } else {
                gip.unallocatedMessage = defaultMsgIPUnavailable
                gip.unallocatedReason = defaultReasonIPUnavailable
            }
        }
```

Reply:
Agree with Tom regarding the usage of conditions.

Reply:
I was thinking of a case where it might be a stale ip.